### PR TITLE
Share device-side LM head across decode and prefill

### PR DESCRIPTION
**Pull request by sjduan**
## Summary

- Run the DeepSeek V4 LM head on device for both packed prefill and decode.
- Share one LM-head projection/transport implementation across both forward paths.
- Support decoupled hidden owners and TP ranks: all eight owners publish selected hidden rows, while only the first four LM-head TP ranks own vocabulary weight shards and compute logits.
- Keep full-vocabulary logits as the output; device-side sampling is outside this PR.
- Reuse the existing MoE distributed windows instead of creating a second HCCL/window transport after the 43-layer forward.

### `models/deepseek/v4/lm_head.py`

- Adds shared owner/TP-decoupled LM-head workers for prefill and decode.
- Selects compact output rows through `num_tokens_per_owner`, `logit_row_indices`, and `num_logit_rows`.
- Owner ranks publish selected hidden rows; TP ranks load all owner rows and compute their vocabulary shards.
- TP ranks route logits shards back to each owner; owners assemble full logits after all shards arrive.
- Uses epoch-aware `AtomicAdd` notifications so the transport can safely reuse synchronization buffers after repeated MoE epochs.

### `models/deepseek/v4/prefill_fwd.py`

- Integrates the shared device LM head after the 43-layer packed prefill forward.
- Reuses the MoE `recv_x_buf`, `arrived_buf`, and `combine_arrived_buf` windows.
- Runs LM-head communication at `LAST_MOE_EPOCH + 1` and allocates only the intermediate TP logits shards.

### `models/deepseek/v4/decode_fwd.py`

- Integrates the same device LM head after the 43-layer decode forward.
- Reuses the same MoE transport buffers and runs LM-head communication at epoch 44.
- Keeps LM-head weights only on the four TP compute ranks.

## MTP Boundary

- Main prefill and main `S=2` speculative decode logits are produced by the device LM head.
- Serving MTP draft prefill/decode helpers remain host-side; draft execution and acceptance semantics are unchanged.

## Validation

- Python static checks and `git diff --check`: PASS.
- Standalone LM head, EP4/TP4, golden compare: PASS.
  - Task: `task_20260716_000218_28707164067`
- Serving-a random-weight 43-layer bring-up, EP8/TP4: PASS.
  - Packed prefill: compile/runtime PASS, total 328.82s.
  - Decode: compile/runtime PASS, total 252.15s.
  - No 507018, HCCL deadlock, timeout, or non-finite output.
- PyPTO real-weight resident+MTP serving: PASS.
  - Task: `task_20260721_182816_173066316561`
  - Devices: `0-7`
  - Model: `/data/models/dsv4-flash-w8a8`
  - Serving launch: `--dp 1 --tp 8`; the device LM-head kernel remains internally TP4.
  - Prompt: `Huawei is`; prompt tokens: 4; completion tokens: 10.
  - MTP accepted/proposed: `4/5` (80%).
  - Request end-to-end time: 730.01s.
  - Output:

    ```text
     a leading global information and communications technology (ICT)
    ```

  - Resident MTP weights uploaded successfully; about 8.49GB of host backing was released.
  - No 507018, HCCL error, timeout, compile/ABI error, or non-finite output.

## Files

- `models/deepseek/v4/lm_head.py`
- `models/deepseek/v4/prefill_fwd.py`
- `models/deepseek/v4/decode_fwd.py`

### CODE REVIEW

**Comment by sjduan** [comment]
Validation update after the shared LM-head arena fix (lib `a734941`, serving `a5b4d30`):

- serving-a `lm_head.py` EP2/TP2 strict logits compare: PASS
- serving-a `lm_head.py` EP8/TP4 strict logits compare: PASS
- serving-a 43-layer `decode_fwd.py` EP2/TP2: PASS
- serving-a 43-layer `decode_fwd.py` EP8/TP4: PASS
- serving-a 43-layer `prefill_fwd.py` EP8/TP4, 128 tokens: PASS
- DeepSeek serving contract suite: 30/30 PASS
- PyPTO cards 8-15 real-weight MTP accuracy task `task_20260721_215819_128002217364`: PASS

Strict completion output for prompt `Huawei is` matched the CI golden exactly: ` a leading global information and communications technology (ICT)`. No 507018, HCCL deadlock, timeout, or non-finite error occurred.

The final transport arrangement reuses the sequential MoE data arena for LM-head hidden/logits payloads, sizes it for the larger payload at module scope (PyPTO JIT cannot parse Python `max()` in the host function), and keeps LM-head completion counters independent from MoE epochs.

**Comment by sjduan** [comment]
Follow-up CI compatibility fix pushed in `1b030b8`.

Root cause: pypto-lib CI checks out upstream pypto-serving `main`, which still calls the legacy `l3_prefill_fwd` / `l3_decode_fwd` ABI. The device-LM-head change had replaced those public symbols, so compilation failed with a missing `logits` argument.

Fix:
- preserve the legacy public host wrappers and scalar `num_tokens` ABI;
- add explicit `l3_prefill_fwd_with_lm_head` / `l3_decode_fwd_with_lm_head` entrypoints for the new serving contract;
- share one inline 43-layer core between both wrappers;
- represent owner token counts as `[EP, 1]`, allowing the host graph to pass a rank-local tensor view without reading tensor data on host.

Validation on serving-a (`openeuler-2403-DS`):
- legacy prefill/decode ABI compile probes: PASS;
- LM-head EP8/TP4 strict logits compare: PASS;
- 43-layer prefill EP8/TP4, 128 active tokens: runtime PASS (332.29s);
- 43-layer decode EP8/TP4, 8 active tokens, start_pos=8192: runtime PASS (227.45s);
- no 507018, HCCL, timeout, or non-finite errors.

**Comment by sjduan** [comment]
Superseding my previous compatibility-layer comment: after review, the PR is intentionally back to the minimal direct contract. We do **not** preserve the old serving ABI in pypto-lib.

Current head: `9872f48`. The compatibility wrappers/duplicate signatures were reverted; the source tree is identical to the previously accuracy-passing `a734941`. `l3_prefill_fwd` and `l3_decode_fwd` are again the sole device-LM-head forward entrypoints, and `num_tokens_per_owner` remains `[EP]`.

Final validation:
- serving-a LM-head EP8/TP4 strict logits compare: PASS;
- serving-a DeepSeek serving tests: 30/30 PASS;
- PyPTO real-weight strict accuracy task `task_20260721_233013_260604217702`: exit 0;
- revision marker: `serving=480a96b lib=9872f48`;
- prompt `Huawei is` -> exact expected text ` a leading global information and communications technology (ICT)`;
- no 507018, HCCL, timeout, non-finite, or ABI errors.

This means lib and serving should be reviewed/landed as one contract update rather than carrying an interim backward-compatibility layer.
